packages/relay/src/lib/services/cacheService/cacheService.ts

```diff
@@ -0,0 +1,187 @@
+
+  /**
+   * The Redis cache used for caching items from requests.
+   *
+   * @private
+   */
+  private readonly sharedCache: ICacheClient;
+
+  /**
+   * The logger used for logging all output from this class.
+   * @private
+   */
+  private readonly logger: Logger;
+
+  /**
+   * The metrics register used for metrics tracking.
+   * @private
+   */
+  private readonly register: Registry;
+
+  /**
+   * Used for reference to the state of REDIS_ENABLED and REDIS_URL env. variables.
+   */
+  private readonly isSharedCacheEnabled: boolean;
+  /**
+   * Represents a caching manager that utilizes various cache implementations based on configuration.
+   * @param {Logger} logger - The logger used for logging all output from this class.
+   * @param {Registry} register - The metrics register used for metrics tracking.
+   */
+  public constructor(logger: Logger, register: Registry) {
+    this.logger = logger;
+    this.register = register;
+
+    this.internalCache = new LocalLRUCache(logger.child({ name: 'localLRUCache' }), register);
+    this.sharedCache = this.internalCache;
+    this.isSharedCacheEnabled = this.isRedisEnabled();
+
+    if (this.isSharedCacheEnabled) {
+      this.sharedCache = new RedisCache(logger.child({ name: 'redisCache' }), register);
+    }
+  }
+
+  /**
+   * Checks whether Redis caching is enabled based on environment variables.
+   * @private
+   * @returns {boolean} Returns true if Redis caching is enabled, otherwise false.
+   */
+  private isRedisEnabled(): boolean {
+    const redisEnabled = process.env.REDIS_ENABLED && process.env.REDIS_ENABLED === 'true';
+    const redisUrlValid = process.env.REDIS_URL && process.env.REDIS_URL !== '';
+
+    if (redisEnabled && redisUrlValid) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Retrieves a cached value associated with the given key.
+   * If the shared cache is enabled and an error occurs while accessing it,
+   * the internal cache is used as a fallback.
+   * @param {string} key - The key associated with the cached value.
+   * @param {string} callingMethod - The name of the method calling the cache.
+   * @param {string} requestIdPrefix - A prefix to include in log messages (optional).
+   * @param {boolean} shared - Whether to use the shared cache (optional, default: false).
+   * @returns {*} The cached value if found, otherwise null.
+   */
+  public async get(key: string, callingMethod: string, requestIdPrefix?: string, shared: boolean = false): Promise<any> {
+    if (shared && this.isSharedCacheEnabled) {
+      try {
+        return await this.sharedCache.get(key, callingMethod, requestIdPrefix);
+      } catch (error) {
+        const redisError = new RedisCacheError(error);
+        this.logger.error(`Error occured while getting the cache from Redis. Fallback to internal cache. Error is: ${redisError.fullError}`);
+      }
+      return this.internalCache.get(key, callingMethod, requestIdPrefix);
```

Comment:
I don't think this fallback should be there. To honour the intent of `shared` we shouldn't manage both. Allow callers to manage that logic
```suggestion
      return null;
```